src/uct/tcp/tcp_ep.c

```diff
@@ -451,6 +451,59 @@ void uct_tcp_ep_pending_queue_dispatch(uct_tcp_ep_t *ep)
                 "ep=%p", ep);
 }
 
+/**
+ * Fill iovec data structure by data provided in uct_iov_t
+ * The function avoids copying IOVs with zero length
+ *
+ * @return Number of elements in io_vec[]
+ */
+static inline size_t
+uct_tcp_ep_iovec_fill_iov(struct iovec *io_vec, const uct_iov_t *iov,
+                          size_t iovcnt, size_t *total_length)
+{
+    size_t iov_it, io_vec_it = 0;
+
+    *total_length = 0;
+
+    for (iov_it = 0; iov_it < iovcnt; ++iov_it) {
+        io_vec[io_vec_it].iov_len = uct_iov_get_length(&iov[iov_it]);
+
+        /* avoid zero length elements in resulted iov_vec */
+        if (io_vec[io_vec_it].iov_len != 0) {
+            io_vec[io_vec_it].iov_base = iov[iov_it].buffer;
+            *total_length += io_vec[io_vec_it].iov_len;
+            ++io_vec_it;
+        }
+    }
+
+    return io_vec_it;
+}
+
+/* Rewinds `iov` array to not consider already consumed buffers */
+static void
+uct_tcp_ep_consume_iov(struct iovec *iov, size_t *iov_cnt, size_t consumed)
+{
+    size_t i;
+
+    if (*iov_cnt == 1) {
+        goto out;
+    }
+
+    for (i = 0; i < *iov_cnt; i++) {
+        if (consumed < iov[i].iov_len) {
+            break;
+        }
+        consumed -= iov[i].iov_len;
+    }
+
+    *iov_cnt -= i;
+    memmove(iov, &iov[i], sizeof(*iov) * *iov_cnt);
```

Comment:
seems like O(n^2) complexity
we should keep "current" pointer (iov_index, iof_offset) instead

Reply:
good catch
done